File: OneQuestionPerDay/718/maximum_length_of_repeated_subarray2.go
package main

func findLength2(A []int, B []int) int {
	m, n := len(A), len(B)
	if m == 0 || n == 0 {
		return 0
	}
	res := 0

	for i := 0; i < m; i++ {
		l := min(m-i, n)
		res = max(res, maxLen(A, i, B, 0, l))
	}
	for i := 0; i < n; i++ {
		l := min(n-i, m)
		res = max(res, maxLen(A, 0, B, i, l))
	}
	return res
}

func min(x, y int) int {
	if x < y {
		return x
	}
	return y
}
func maxLen(a []int, i int, b []int, j int, m int) int {
	res := 0
	k := 0
	for l := 0; l < m; l++ {
		if a[i+l] == b[j+l] {
			k++
			res = max(res, k)
		} else {
			k = 0
		}
	}
	return res
}

File: OneQuestionPerDay/718/maximum_length_of_repeated_subarray2.py
from typing import List


class Solution:
    def findLength(self, A: List[int], B: List[int]) -> int:
        def maxLength(addA: int, addB: int, length: int) -> int:
            ret = k = 0
            for i in range(length):
                if A[addA + i] == B[addB + i]:
                    k += 1
                    ret = max(ret, k)
                else:
                    k = 0
            return ret

        n, m = len(A), len(B)
        ret = 0
        for i in range(n):
            length = min(m, n - i)
            ret = max(ret, maxLength(i, 0, length))
        for i in range(m):
            length = min(n, m - i)
            ret = max(ret, maxLength(0, i, length))
        return ret

    def test(self):
        a = [1, 2, 3, 2, 1]
        b = [3, 2, 1, 4, 7]
        ans = self.findLength(a, b)
        print(ans)


if __name__ == "__main__":
    Solution().test()
